lib/middleware/request-handler.ts

```diff
@@ -290,6 +249,196 @@ export function onSubscribe(req: Interface.IOurRequest,
+        return next(new restify.BadRequestError('No active subscriptions.'));
+    }
+
+    var subscriptions: Subscription[] = [];
+    // If no correlation Id specified,
+    // the default behavior is to unsubscribe all active subscriptions
+    if (!req.body) {
+        subscriptions = req.apiSession.activeSubscriptions.values();
+    } else {
+        // If we do receive req.body object, first check if it is valid(empty list is INVALID)
+        if (!_.has(req.body, 'correlationIds') ||
+            !_.isArray(req.body.correlationIds) ||
+            !req.body.correlationIds.length) {
+            req.log.debug('Invalid unsubscribe data.');
+            return next(new restify.InvalidArgumentError('Invalid unsubscribe data.'));
+        }
+        // Next, validate all correlation Ids
+        // Will error if any invalid correlation Id received
+        var isAllValid = true;
+        _.uniq(req.body.correlationIds).forEach((cid: number): boolean => {
+            if (req.apiSession.activeSubscriptions.has(cid)) {
+                subscriptions.push(req.apiSession.activeSubscriptions.get(cid));
+            } else {
+                isAllValid = false;
+                req.log.debug('Invalid correlation Id ' + cid + ' received.');
+                return false;
+            }
+        });
+        if (!isAllValid) {
+            return next(new restify.InvalidArgumentError('Invalid correlation id.'));
+        }
+    }
+
+    try {
+        var result: Object[] = [];
+        req.blpSession.unsubscribe(subscriptions);
+        subscriptions.forEach((sub: Subscription): void => {
+            sub.removeAllListeners();
+            if (!sub.buffer.isEmpty()) {
+                result.push(sub.buffer.startNewBuffer());
+            }
+            req.apiSession.activeSubscriptions.delete(sub.correlationId);
+            req.apiSession.receivedSubscriptions.delete(sub.correlationId);
+        });
+        req.log.debug({ activeSubscriptions:
+                       req.apiSession.activeSubscriptions.size },
+                      'Unsubscribed.');
+
+        // Reset poll Id to null if all subscriptions get unsubscribed
+        if (!req.apiSession.receivedSubscriptions.size) {
+            req.apiSession.lastPollId = req.apiSession.lastSuccessPollId = null;
+        }
+
+        res.sendChunk(result);
+        res.sendEnd(0, 'Unsubscribe Successfully');
+        return next();
+    } catch (err) {
+        req.log.error(err, 'Unsubscription error');
+        return next(new restify.InternalError(err.message));
+    }
+}
+
+function onGetToken(req: Interface.IOurRequest,
+                    res: Interface.IOurResponse,
+                    next: restify.Next): void
+{
+    var callback = getRequestCallback(req, res, next);
+    var uri = '//blp/apiauth';
+    var requestName = 'AuthorizationTokenRequest';
+    req.blpSession.request(uri, requestName, req.body, undefined /* identity */, callback);
```

Comment:
I understand what is happening here, but it seems misleading having this function take `next` without having it visible that it gets called. I would advocate for calling next explicitly here or reworking the function signature and how everything fits together.